portal-impl/src/com/liferay/portal/service/impl/UserGroupLocalServiceImpl.java

```diff
@@ -1286,31 +1284,32 @@ protected void reindexUserGroup(UserGroup userGroup)
 	protected void reindexUsers(List<UserGroup> userGroups)
 		throws PortalException {
 
-		Stream<UserGroup> stream1 = userGroups.stream();
+		Map<Long, List<Long>> map = new HashMap<>();
 
-		Map<Long, List<UserGroup>> map = stream1.collect(
-			Collectors.groupingBy(UserGroup::getCompanyId));
-
-		for (Map.Entry<Long, List<UserGroup>> entry : map.entrySet()) {
-			long companyId = entry.getKey();
+		for (UserGroup userGroup : userGroups) {
+			List<Long> userGroupIds = map.computeIfAbsent(
+				userGroup.getCompanyId(), key -> new ArrayList<>());
 
-			List<UserGroup> list = entry.getValue();
+			userGroupIds.add(userGroup.getUserGroupId());
+		}
 
-			Stream<UserGroup> stream2 = list.stream();
+		for (Map.Entry<Long, List<Long>> entry : map.entrySet()) {
+			long companyId = entry.getKey();
 
-			final long[] userGroupIds = stream2.mapToLong(
-				UserGroup::getUserGroupId
-			).toArray();
+			List<Long> userGroupIds = entry.getValue();
 
 			TransactionCommitCallbackUtil.registerCallback(
 				() -> {
-					LongStream longStream = Arrays.stream(userGroupIds);
+					Set<Long> userIdSet = new HashSet<>();
+
+					for (long userGroupId : userGroupIds) {
+						for (long userId : getUserPrimaryKeys(userGroupId)) {
+							userIdSet.add(userId);
+						}
+					}
 
-					long[] userIds = longStream.flatMap(
-						userGroupId -> Arrays.stream(
-							getUserPrimaryKeys(userGroupId))
-					).distinct(
-					).toArray();
+					long[] userIds = TransformUtil.transformToLongArray(
```

Comment:
@slnn Please try to analyze the business logic, simplify this whole method and reduce iterations.